ENH: Add optional argument index to pd.melt to maintain index values

Rik-de-Kort opened the pull request:
Finishing up a stale PR idea: #28859 and #17459

 Has some tests and better code.
I think it's fair to duplicate the index values and not bend over backwards to maintain uniqueness like in previous iterations.

Apologies for the mess, it was a quick job and I didn't want to spend an hour fiddling with the commits.

Finally, I deleted some ignore type comments for mypy because the commits weren't going on my system. Is there some other fix for that? Other than that I think it's good to go.

- [X] closes #17440
- [X] tests added / passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry
- [x] Reconsider API design
- [x] Add usage examples in whatsnew, docstring, and reshaping.rst.

Rik-de-Kort committed:
initial xlsb support

Rik-de-Kort committed:
Import order fix for CI pass

Rik-de-Kort committed:
Initial tests

Rik-de-Kort committed:
style fixes

Rik-de-Kort committed:
documentation

Rik-de-Kort committed:
forgot place to document

Rik-de-Kort committed:
Fixed test issue with XLRDError

Rik-de-Kort committed:
Fix for unnamed column issue

Rik-de-Kort committed:
style fix

Rik-de-Kort committed:
line up with upstream master

Rik-de-Kort committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

Rik-de-Kort committed:
Fix broken xlrd test

Rik-de-Kort committed:
get docs to build

Rik-de-Kort committed:
Remove warning filter

Rik-de-Kort committed:
Merge branch 'master' of https://github.com/Rik-de-Kort/pandas

Rik-de-Kort committed:
extended description update

Rik-de-Kort committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

Rik-de-Kort committed:
Xlsb options instead of odf options

Rik-de-Kort committed:
Add reference in whatsnew to docs

Rik-de-Kort committed:
Make pyxlsb show up in install.rst and show_versions

Rik-de-Kort committed:
Add pyxlsb to ci builds

Rik-de-Kort committed:
environment.yml update

Rik-de-Kort committed:
Merge upstream master

Rik-de-Kort committed:
One update to environment.yml too many

Rik-de-Kort committed:
Trying to fix build

Rik-de-Kort committed:
Merge upstream

Rik-de-Kort committed:
Added issue number

Rik-de-Kort committed:
Updated to use .rows(sparse=False) for future compat

Rik-de-Kort committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

Rik-de-Kort committed:
xfails in test_readers.py

Rik-de-Kort committed:
xfail url loads

Rik-de-Kort committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

Rik-de-Kort committed:
Update min version

Rik-de-Kort committed:
test xfailing for the right reason

Rik-de-Kort committed:
xfail unnecessary due to consistency check only

Rik-de-Kort committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

Rik-de-Kort committed:
Merge branch 'master' of https://github.com/Rik-de-Kort/pandas

Rik-de-Kort committed:
Initial fix over from original repo.

Rik-de-Kort committed:
Proper implementation

Rik-de-Kort committed:
Multiindex support

Rik-de-Kort committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

Rik-de-Kort committed:
Updated docstring

Rik-de-Kort committed:
Whatsnew entry

Rik-de-Kort committed:
Fix mypy error

simonjayhawkins commented:
Thanks @Rik-de-Kort for the PR.

> Finishing up a stale PR idea, here: #28859

The original work is #17459

Can you review that PR and also address any comments raised there that are relevant to this PR. e.g.

@TomAugspurger wrote https://github.com/pandas-dev/pandas/pull/17459#issuecomment-327818758

> For the docs, it'll need a whatsnew note in doc/source/whatsnew/v0.21.0.txt, prose docs in doc/source/reshaping.rst. It'd also be nice to have an example in the docstring.

can you update reshaping.rst and add examples to docstring. The release note should probably also include examples demonstrating the new functionality.

Rik-de-Kort committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

Rik-de-Kort committed:
renamed kwarg, docstring

pep8speaks commented:
Hello @Rik-de-Kort! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/frame.py`](https://github.com/pandas-dev/pandas/blob/0455f6478f77544d4e157cba7bc52ec7f5d5edcc/pandas/core/frame.py):

> [Line 6713:1](https://github.com/pandas-dev/pandas/blob/0455f6478f77544d4e157cba7bc52ec7f5d5edcc/pandas/core/frame.py#L6713): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 6747:1](https://github.com/pandas-dev/pandas/blob/0455f6478f77544d4e157cba7bc52ec7f5d5edcc/pandas/core/frame.py#L6747): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

Rik-de-Kort committed:
updated reshaping.rst

Rik-de-Kort committed:
PEP8 issue

Rik-de-Kort committed:
PEP8 issue

Rik-de-Kort committed:
Updated whatsnew

Rik-de-Kort commented:
Thanks for commenting, @simonjayhawkins I couldn't find that PR when I was looking for it!

* I have renamed the kwarg to `index` rather than `keep_index` as suggested by @jreback.
* I have added an example to the docstring and reshaping.rst.

I don't think the whatsnew needs more prose about the change. It's use is straightforward: as opposed to the previous PR's it does the simplest thing imaginable, namely keep the original index values around in their original form.  
Speaking from an end user perspective (I use pandas a lot in my work), I think this is what I would like the API to be, but of course I'm open to discussion.  
One alternative would be to create a unique index using the old index values and `id_vars` as in the previous PRs. However, that is rather unwieldy in terms of code (it's bad enough to have to handle the case for the multiindex as it is), and also not very consistent with the behaviour when `index=False` because that keeps `id_vars` around as a column rather than as part of the index.

Rik-de-Kort committed:
trailing whitespace in reshaping.rst

jreback commented on doc/source/user_guide/reshaping.rst:
```diff
@@ -296,6 +296,18 @@ For instance,
    cheese.melt(id_vars=['first', 'last'], var_name='quantity')
 
+Optionally, original index values can be kept around using the ``index`` parameter.
+
```
add a version added tag here

you can make this a subsection as well

Rik-de-Kort replied:
Done on both accounts.

jreback commented on doc/source/user_guide/reshaping.rst:
```diff
@@ -296,6 +296,18 @@ For instance,
+                          'height': [5.5, 6.0],
+                          'weight': [130, 150]})
+   cheese
+   cheese.melt(id_vars=['first', 'last'], index=True)
```
put the actual operations in a separate i python block (they render better)

Rik-de-Kort replied:
I have done this, but the rest of the documentation doesn't do it, and I'm not sure I like the look. Maybe I misunderstood?

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -99,7 +99,7 @@ Other enhancements
   (:issue:`33196`)
 - :meth:`~pandas.core.groupby.GroupBy.transform` has gained ``engine`` and ``engine_kwargs`` arguments that supports executing functions with ``Numba`` (:issue:`32854`)
--
+- :meth: `melt` has gained an ``index`` argument that prevents the method from dropping the index.
```
add the issue number

the ref to melt has an extra space (after the :)

Rik-de-Kort replied:
Done

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -99,7 +99,7 @@ Other enhancements
   (:issue:`33196`)
 - :meth:`~pandas.core.groupby.GroupBy.transform` has gained ``engine`` and ``engine_kwargs`` arguments that supports executing functions with ``Numba`` (:issue:`32854`)
--
+- :meth: `melt` has gained an ``index`` argument that prevents the method from dropping the index.
```
say what the default is

jreback commented on pandas/core/frame.py:
```diff
@@ -6658,6 +6658,9 @@ def unstack(self, level=-1, fill_value=None):
         Name to use for the 'value' column.
     col_level : int or str, optional
         If columns are a MultiIndex then use this level to melt.
+    index : boolean, optional, default False
```
let’s rename to keep_index while longer is more descriptive

jreback commented on pandas/core/frame.py:
```diff
@@ -6708,6 +6711,17 @@ def unstack(self, level=-1, fill_value=None):
 
+    Original index values can be kept around:
+
+    >>> %(caller)sid_vars=['A'], value_vars=['B', 'C'], index=True)
```
not sure what u r doing here

Rik-de-Kort replied:
Adding an example to the docstring. Passes docstring verification.

jreback commented on pandas/core/reshape/melt.py:
```diff
@@ -115,7 +116,17 @@ def melt(
+    if index:
+        new_index = np.tile(frame.index, K)
+
+        if isinstance(frame.index, MultiIndex):  # Index of tuples otherwise
```
you might be able to use .from_arrays instead

Rik-de-Kort replied:
It seems the natural "arrangement" for MultiIndex values is in tuples form. There is the `codes`-property but that doesn't work because it's not the original index values. I think using from_tuples is preferable in this case.

jreback commented on pandas/tests/reshape/test_melt.py:
```diff
@@ -333,6 +333,29 @@ def test_melt_mixed_int_str_value_vars(self):
         tm.assert_frame_equal(result, expected)
 
+    def test_keep_index(self):
+        df = DataFrame({"foo": [0], "bar": [1]}, index=["first"])
```
add the issue number as. comment

Rik-de-Kort committed:
Resolve comments

Rik-de-Kort committed:
Maybe this will fix doc linting?

simonjayhawkins commented on pandas/core/reshape/melt.py:
```diff
@@ -115,7 +116,17 @@ def melt(
+    result = frame._constructor(mdata, columns=mcolumns)
+
+    if keep_index:
+        new_index = np.tile(frame.index, K)
```
Index types (e.g. Categorical) and index names are not retained?

Rik-de-Kort replied:
Good observation, added.

Rik-de-Kort committed:
Names and types

Rik-de-Kort committed:
Merge upstream master

Rik-de-Kort committed:
Merging master, again

TomAugspurger commented on doc/source/user_guide/reshaping.rst:
```diff
@@ -312,6 +313,27 @@ user-friendly.
+
+Keeping the index around
+~~~~~~~~~~~~~~~~~~~~~~~~
+When transforming a DataFrame using :func:`~pandas.melt`, original index values can be kept around by setting the ``keep_index`` parameter to True. This will however duplicate the index values.
```
```suggestion

When transforming a DataFrame using :func:`~pandas.melt`, original the index values can be kept around by setting the ``keep_index`` parameter to True. This will however duplicate the index values.
```

TomAugspurger commented on pandas/core/frame.py:
```diff
@@ -6811,6 +6811,9 @@ def unstack(self, level=-1, fill_value=None):
     col_level : int or str, optional
         If columns are a MultiIndex then use this level to melt.
+    keep_index : bool, optional, default False
+            If True, the original index is reused.
+            There is no uniqueness guarantee.
```
```suggestion
        If True, the original index is retained. Index labels will be repeated
        as necessary.
```

TomAugspurger reviewed (approved):
Looks nice.

The repeated values from `keep_index=True` naturally makes me think that we should form a MultIndex out of the `id_vars` and the original index, but that can be a separate issue.

Rik-de-Kort committed:
Update doc/source/user_guide/reshaping.rst

Co-Authored-By: Tom Augspurger <TomAugspurger@users.noreply.github.com>

Rik-de-Kort committed:
Update pandas/core/frame.py

Co-Authored-By: Tom Augspurger <TomAugspurger@users.noreply.github.com>

Rik-de-Kort committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

Rik-de-Kort committed:
Merge branch 'master' of https://github.com/Rik-de-Kort/pandas

Rik-de-Kort commented:
Some issues with the CI mysteriously failing. Seems resolved now.

@jreback @simonjayhawkins if you have time, could you please review? Of course understandable if priorities lie elsewhere :)

jreback commented on doc/source/user_guide/reshaping.rst:
```diff
@@ -312,6 +313,28 @@ user-friendly.
+Keeping the index around
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+When transforming a DataFrame using :func:`~pandas.melt`, original the index values can be kept around by setting the ``keep_index`` parameter to True. This will however duplicate the index values.
```
add a versionadded 1.1

jreback replied:
move from below

jreback commented on doc/source/user_guide/reshaping.rst:
```diff
@@ -312,6 +313,28 @@ user-friendly.
+
+.. ipython:: python
+
+   index = pd.MultiIndex.from_tuples([('person', 'first'), ('person', 'second')])
```
this should be after the melt example above.

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -146,11 +146,12 @@ Other enhancements
   This can be used to set a custom compression level, e.g.,
   ``df.to_csv(path, compression={'method': 'gzip', 'compresslevel': 1}``
   (:issue:`33196`)
+- :meth:`~pandas.core.groupby.GroupBy.transform` has gained ``engine`` and ``engine_kwargs`` arguments that supports executing functions with ``Numba`` (:issue:`32854`)
```
why is this changed? (the transform note)

jreback commented on pandas/core/frame.py:
```diff
@@ -6811,6 +6811,9 @@ def unstack(self, level=-1, fill_value=None):
     col_level : int or str, optional
         If columns are a MultiIndex then use this level to melt.
+    keep_index : bool, optional, default False
+        If True, the original index is retained. Index labels will be repeated
```
needs a versionadded 1.1

jreback commented on pandas/core/reshape/melt.py:
```diff
@@ -115,7 +116,19 @@ def melt(
+        else:
+            new_index = Index(new_index, dtype=frame.index.dtype, name=frame.index.name)
+
+        result = result.set_index(new_index)
```
result.index = new_index is enough here (its already a copy)

jreback commented on pandas/tests/reshape/test_melt.py:
```diff
@@ -333,6 +333,42 @@ def test_melt_mixed_int_str_value_vars(self):
+        result = melt(df, keep_index=True)
+
+        expected_index = pd.Index(["foo", "bar"] * 2, dtype="category", name="baz")
+        tm.assert_index_equal(result.index, expected_index)
```
use assert_frame_equal and construct the expected

jreback commented on pandas/core/frame.py:
```diff
@@ -6811,6 +6811,9 @@ def unstack(self, level=-1, fill_value=None):
         Name to use for the 'value' column.
     col_level : int or str, optional
         If columns are a MultiIndex then use this level to melt.
+    keep_index : bool, optional, default False
```
this needs to be called ignore_index as we use this everywhere else, default True i think (as that is what we do now)

Rik-de-Kort committed:
Merge upstream master

Rik-de-Kort committed:
Reviewer comments

Rik-de-Kort committed:
transform note?

Rik-de-Kort committed:
Test full frame

Rik-de-Kort committed:
Linting

simonjayhawkins commented on doc/source/user_guide/reshaping.rst:
```diff
@@ -312,6 +331,8 @@ user-friendly.
   pd.wide_to_long(dft, ["A", "B"], i="id", j="year")
 
+
+
```
why are these blank lines added?

simonjayhawkins commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -220,6 +220,8 @@ Other enhancements
   This can be used to set a custom compression level, e.g.,
   ``df.to_csv(path, compression={'method': 'gzip', 'compresslevel': 1}``
   (:issue:`33196`)
+- :meth:`~pandas.core.groupby.GroupBy.transform` and :meth:`~pandas.core.groupby.GroupBy.aggregate` has gained ``engine`` and ``engine_kwargs`` arguments that supports executing functions with ``Numba`` (:issue:`32854`, :issue:`33388`)
```
why is this added? from resolving a merge conflict maybe?

Rik-de-Kort replied:
Yeah that's why I put it there, but it seems to be unnecessary.

Rik-de-Kort committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

Rik-de-Kort committed:
Documentation fixes

Rik-de-Kort commented:
Does the CI always have this many issues?

simonjayhawkins commented on pandas/core/frame.py:
```diff
@@ -6921,6 +6921,11 @@ def unstack(self, level=-1, fill_value=None):
     col_level : int or str, optional
         If columns are a MultiIndex then use this level to melt.
+    ignore_index : bool, optional, default True
+        If False, original index is ignored. If True, the original index is retained.
```
is this correct? if ignore_index=False, the index is ignored?

Rik-de-Kort replied:
Oh my word... Thanks for being so thorough!

Rik-de-Kort committed:
Fixed documentation

Rik-de-Kort committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

Rik-de-Kort committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

Rik-de-Kort committed:
Merge branch 'master' of https://github.com/Rik-de-Kort/pandas

Rik-de-Kort committed:
Fixed docs (Hopefully)

simonjayhawkins commented on pandas/core/frame.py:
```diff
@@ -7008,7 +7008,6 @@ def unstack(self, level=-1, fill_value=None):
              b   4.0
         dtype: float64
 
-        >>> s.unstack(level=-1)
```
@Rik-de-Kort I don't think you intended to remove this.

 the doc error is `/home/runner/work/pandas/pandas/pandas/core/frame.py:docstring of pandas.DataFrame.unstack:55: WARNING: Block quote ends without a blank line; unexpected unindent.`

so could be related.

Rik-de-Kort replied:
ahhh thanks! all checks passing now, finally!

Rik-de-Kort committed:
Merge upstream master

Rik-de-Kort committed:
Hopefully fix documentation bug

Rik-de-Kort committed:
Fix typing error

simonjayhawkins commented on doc/source/user_guide/reshaping.rst:
```diff
@@ -296,6 +296,25 @@ For instance,
    cheese.melt(id_vars=['first', 'last'])
    cheese.melt(id_vars=['first', 'last'], var_name='quantity')
 
+When transforming a DataFrame using :func:`~pandas.melt`, original the index values can be kept around by setting the ``ignore_index`` parameter to ``False``. This will however duplicate the index values.
```
```suggestion
When transforming a DataFrame using :func:`~pandas.melt`, the original index values can be kept around by setting the ``ignore_index`` parameter to ``False``. This will however duplicate the index values.
```

simonjayhawkins commented on doc/source/user_guide/reshaping.rst:
```diff
@@ -296,6 +296,25 @@ For instance,
+                         index=index)
+   cheese
+
+.. ipython:: python
```
no need to repeat the directive here, as it creates an extra box and splits the example

simonjayhawkins commented on doc/source/user_guide/reshaping.rst:
```diff
@@ -296,6 +296,25 @@ For instance,
+
+.. ipython:: python
+
+   index = pd.MultiIndex.from_tuples([('person', 'first'), ('person', 'second')])
```
this example is a bit confusing with first and second since first and last are columns and used as id_vars latter in the example. maybe just 1 and 2 or A and B?

simonjayhawkins reviewed (commented):
Thanks @Rik-de-Kort almost there. generally lgtm. just a couple more comments.

jreback commented on doc/source/user_guide/reshaping.rst:
```diff
@@ -296,6 +296,25 @@ For instance,
+
+   cheese.melt(id_vars=['first', 'last'], ignore_index=False)
+
+.. versionadded:: 1.1.0
```
can you put this at the beginning of the section

jreback commented on pandas/core/frame.py:
```diff
@@ -7086,6 +7086,7 @@ def melt(
         var_name=None,
         value_name="value",
         col_level=None,
+        ignore_index=True,
```
can you type ( i know nothing else is, but at least this is a start)

Rik-de-Kort replied:
Of course!

jreback commented on pandas/core/reshape/melt.py:
```diff
@@ -121,7 +121,19 @@ def melt(
+    result = frame._constructor(mdata, columns=mcolumns)
+
+    if not ignore_index:
+        new_index = np.tile(frame.index, K)
```
both MI and Index already have a .repeat() method, I think we could add a .tile() method to make this easier. (or just use repeat)

Rik-de-Kort replied:
`Index(["foo", "bar"]).repeat(2)` yields `Index(['foo', 'foo', 'bar', 'bar'], dtype='object')`, where as `np.tile(["foo", "bar"])` yields `array(['foo', 'bar', 'foo', 'bar', 'foo', 'bar'], dtype=object)`. The latter corresponds to the layout used in `melt` so it's very not trivial to use repeat instead of tile.

I tried having a look at implementing `tile` on indices but then I would also have to do it for multiindices and document it and tests, and argument validation which I've never even looked at before and I think it's a big hassle that I will not undertake.

jreback replied:
ok fair enough

simonjayhawkins replied:
we already have _tile_compat in pandas\core\reshape\util.py. This may allow futher simplification here.

TomAugspurger replied:
This converts to object dtype. Can you use `pandas.core.reshape.util._tile_compat`?:

TomAugspurger replied:
You should be able to remove the next section as well, since you won't be converting to an ndarray.

Rik-de-Kort replied:
Thanks, that simplifies the code a lot.

Rik-de-Kort committed:
Apply suggestions from code review

Co-authored-by: Simon Hawkins <simonjayhawkins@gmail.com>

Rik-de-Kort committed:
Doc review:

Rik-de-Kort committed:
Type!

jreback commented on doc/source/user_guide/reshaping.rst:
```diff
@@ -296,6 +296,21 @@ For instance,
+                          'weight': [130, 150]},
+                         index=index)
+   cheese
+   cheese.melt(id_vars=['first', 'last'], ignore_index=False)
```
can you add an example w/o passing ignore_index and indicate that the default is true (unless we already have that)?

e.g. put *another* ipython block starting L312 and then put the False case after

Rik-de-Kort replied:
@jreback it is green therefore ping

jreback reviewed (changes requested):
lgtm. a doc comment, pls ping on green.

Rik-de-Kort committed:
Added example for difference

Rik-de-Kort committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas

Rik-de-Kort committed:
Linting failure?

Rik-de-Kort commented:
@jreback I though I gave you a ping, but I don't see it, so here it is!

Rik-de-Kort committed:
TomAugspurger suggestion

Rik-de-Kort committed:
Trailing whitespace...

jreback commented:
thanks @Rik-de-Kort

Rik-de-Kort commented:
You're welcome
